gtk4/src/subclass/accessible_text.rs

```diff
@@ -0,0 +1,533 @@
+
+    let r_ranges = imp.selection();
+    let n: usize = r_ranges.len();
+    *n_ranges = n;
+
+    if n == 0 {
+        false
+    } else {
+        *ranges = r_ranges.to_glib_container().0;
+
+        true
+    }
+    .into_glib()
+}
+
+unsafe extern "C" fn accessible_text_get_attributes<T: AccessibleTextImpl>(
+    accessible_text: *mut ffi::GtkAccessibleText,
+    offset: libc::c_uint,
+    n_ranges: *mut libc::size_t,
+    ranges: *mut *mut ffi::GtkAccessibleTextRange,
+    attribute_names: *mut *mut *mut libc::c_char,
+    attribute_values: *mut *mut *mut libc::c_char,
+) -> glib::ffi::gboolean {
+    let instance = &*(accessible_text as *mut T::Instance);
+    let imp = instance.imp();
+
+    let attrs = imp.attributes(offset);
+    let n: usize = attrs.len();
+    *n_ranges = n;
+
+    if n == 0 {
+        *attribute_names = std::ptr::null_mut();
+        *attribute_values = std::ptr::null_mut();
+
+        false
+    } else {
+        let mut c_ranges = glib::Slice::with_capacity(attrs.len());
+        let mut c_names = glib::StrV::with_capacity(attrs.len());
+        let mut c_values = glib::StrV::with_capacity(attrs.len());
+
+        for (range, name, value) in attrs {
+            c_ranges.push(range);
+            c_names.push(name);
+            c_values.push(value);
+        }
+
+        *ranges = c_ranges.to_glib_container().0;
+        *attribute_names = c_names.into_glib_ptr();
+        *attribute_values = c_values.into_glib_ptr();
+
+        true
+    }
+    .into_glib()
+}
+
+unsafe extern "C" fn accessible_text_get_default_attributes<T: AccessibleTextImpl>(
+    accessible_text: *mut ffi::GtkAccessibleText,
+    attribute_names: *mut *mut *mut libc::c_char,
+    attribute_values: *mut *mut *mut libc::c_char,
+) {
+    let instance = &*(accessible_text as *mut T::Instance);
+    let imp = instance.imp();
+
+    let attrs = imp.default_attributes();
+
+    if attrs.is_empty() {
+        *attribute_names = std::ptr::null_mut();
+        *attribute_values = std::ptr::null_mut();
+    } else {
+        let mut c_names = glib::StrV::with_capacity(attrs.len());
+        let mut c_values = glib::StrV::with_capacity(attrs.len());
+
+        for (name, value) in attrs {
+            c_names.push(name);
+            c_values.push(value);
+        }
+
+        *attribute_names = c_names.into_glib_ptr();
+        *attribute_values = c_values.into_glib_ptr();
+    }
+}
+
+#[cfg(feature = "v4_16")]
+#[cfg_attr(docsrs, doc(cfg(feature = "v4_16")))]
+unsafe extern "C" fn accessible_text_get_extents<T: AccessibleTextImpl>(
+    accessible_text: *mut ffi::GtkAccessibleText,
+    start: libc::c_uint,
+    end: libc::c_uint,
+    extents: *mut graphene::ffi::graphene_rect_t,
+) -> glib::ffi::gboolean {
+    let instance = &*(accessible_text as *mut T::Instance);
+    let imp = instance.imp();
+
+    let rect = imp.extents(start, end);
+
+    if let Some(rect) = rect {
+        std::ptr::copy_nonoverlapping(rect.into_glib_ptr(), extents, 1);
```

Comment:
```suggestion
        *extents = *rect.as_ptr();
```

or so? `into_glib_ptr()` does a new heap allocation that you leak here